### PR TITLE
core: support saving and loading error artifacts

### LINKED ISSUE #4984: -G/-A doesn't preserve gatherer errors
Internally they're real Errors, so they get serialized as `{}` :( This will obviously be an issue as errors won't show up in the `-A` stage. Presumably there will still be errors when audits try to use those artifacts, just not the ones you'd want to tell what's going on.

It would be nice to continue to use real Errors, so we should support saving and restoring errors in `asset-saver` if possible.

See [skipped test TODO in `runner-test.js`](https://github.com/GoogleChrome/lighthouse/blob/fff50f2bfe59baa5b0083424600b5c4e9a6028c3/lighthouse-core/test/runner-test.js#L276) which needs this in order to pass.

### PR DESCRIPTION
**Pull request by brendankenny**
fixes #4984

> Internally [gatherer errors are] real Errors, so they get serialized as `{}` :( This will obviously be an issue as errors won't show up in the `-A` stage. Presumably there will still be errors when audits try to use those artifacts, just not the ones you'd want to tell what's going on.

> It would be nice to continue to use real Errors, so we should support saving and restoring errors in asset-saver if possible.

This is done through using a `JSON.stringify()` [replacer](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify#The_replacer_parameter) and a `JSON.parse()` [reviver](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#Using_the_reviver_parameter).

It would be possible to do the stringify part by adding a `toJSON` to `LHError`, but we still use regular `Error` in a bunch of places (and allow regular exceptions to bubble up as gatherer errors), so we want to serialize those as well. I don't feel good about changing the global `Error` prototype, though, so here we are :) The implementations are basically the same, though.

Serialization is almost always gross (luckily this is strictly limited to errors and artifacts, not errors in general), so I'm happy to adapt if folks have some cleaner ideas.

### CODE REVIEW

File: lighthouse-core/lib/lh-error.js [hunk 1]
```diff
@@ -96,6 +102,74 @@ class LighthouseError extends Error {
...
+      const code = err.code;
+      return {
+        sentinel: ERROR_SENTINEL,
+        message,
+        code,
+        stack,
+      };
+    }
+
+    throw new Error('Invalid value for LHError stringification');
```
Comment (brendankenny) [inline lighthouse-core/lib/lh-error.js]: this could pass through like `parseReviver`, but then we wouldn't have a foolproof way to have tsc type check the constructed error substitutes, so this seemed ok

File: lighthouse-core/lib/lh-error.js [hunk 2]
```diff
@@ -227,7 +301,7 @@ const ERRORS = {
   },
 
   /* Protocol timeout failures
-   * Requires an additional `icuProtocolMethod` field for translation.
+   * Requires an additional `protocolMethod` field for translation.
```
Comment (brendankenny) [inline lighthouse-core/lib/lh-error.js]: driveby, I assume changed at some point and missed

File: lighthouse-core/lib/lh-error.js [hunk 3]
```diff
@@ -56,10 +55,17 @@ const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
...
+/**
+ * @typedef {{sentinel: '__LighthouseErrorSentinel', code: string, stack?: string, [p: string]: string|undefined}} SerializedLighthouseError
+ * @typedef {{sentinel: '__ErrorSentinel', message: string, code?: string, stack?: string}} SerializedBaseError
+ */
+
 class LighthouseError extends Error {
   /**
    * @param {LighthouseErrorDefinition} errorDefinition
-   * @param {Record<string, string|boolean|undefined>=} properties
+   * @param {Record<string, string|undefined>=} properties
```
Comment (brendankenny) [inline lighthouse-core/lib/lh-error.js]: I added `boolean` here in #6014 but have no memory of why, and it doesn't appear to do anything in that PR or in the current codebase, so I'll assume it was from an early edit of that PR and it slipped through at the end unnoticed :)

**Comment by connorjclark** [comment]
would be nice to have this for [LR](https://github.com/GoogleChrome/lighthouse/blob/master/clients/lightrider/lightrider-entry.js#L61) too.

File: lighthouse-core/lib/lh-error.js [hunk 4]
```diff
@@ -56,10 +55,17 @@ const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
  * @property {boolean} [lhrRuntimeError] True if it should appear in the top-level LHR.runtimeError property.
  */
 
+const LHERROR_SENTINEL = '__LighthouseErrorSentinel';
```
Comment (paulirish) [inline lighthouse-core/lib/lh-error.js]: is this an occasion to use `Symbol()`??

Comment (paulirish) [reply lighthouse-core/lib/lh-error.js]: ah damn it. nope. won't be the same across separate executions.

File: lighthouse-core/lib/lh-error.js [hunk 5]
```diff
@@ -96,6 +103,74 @@ class LighthouseError extends Error {
     const error = new Error(`Protocol error ${errMsg}`);
     return Object.assign(error, {protocolMethod: method, protocolError: protocolError.message});
   }
+
+  /**
+   * A JSON.stringify replacer to serialize LHErrors and (as a fallback) Errors.
+   * Returns a simplified version of the error object that can be reconstituted
+   * as a copy of the original error at parse time.
+   * @param {unknown} err
```
Comment (paulirish) [inline lighthouse-core/lib/lh-error.js]: why not a Error|LighthouseError ? curious.

Comment (brendankenny) [reply lighthouse-core/lib/lh-error.js]: > why not a Error|LighthouseError ? curious.

that's probably better. I was originally writing it more generalized but it wasn't helpful at all

File: lighthouse-core/lib/lh-error.js [hunk 6]
```diff
@@ -96,6 +103,74 @@ class LighthouseError extends Error {
...
+
+      return {
+        sentinel: LHERROR_SENTINEL,
+        code,
+        stack,
+        ...properties,
+      };
+    }
+
+    // We still have some errors that haven't moved to be LHErrors, so serialize them as well.
```
Comment (patrickhulce) [inline lighthouse-core/lib/lh-error.js]: any error that happens unexpectedly will always be a non-LHError, so it's a condition that's likely to exist forever :)

```suggestion
    // Unexpected errors won't be `LHError`, but we still want to serialize them as well.
```

Comment (patrickhulce) [reply lighthouse-core/lib/lh-error.js]: we could leave the line in if we want to also nudge folks to look into the still expected ones that aren't LHError yet?

File: lighthouse-core/lib/lh-error.js [hunk 7]
```diff
@@ -96,6 +103,74 @@ class LighthouseError extends Error {
...
+   * LHError, the error is recreated as the original object. Otherwise, the
+   * value is passed through unchanged.
+   * @param {string} key
+   * @param {any} possibleError
+   * @return {any}
+   */
+  static parseReviver(key, possibleError) {
+    if (typeof possibleError === 'object' && possibleError !== null) {
+      if (possibleError.sentinel === LHERROR_SENTINEL) {
+        // eslint-disable-next-line no-unused-vars
```
Comment (patrickhulce) [inline lighthouse-core/lib/lh-error.js]:
```suggestion
        // eslint-disable-next-line no-unused-vars - unpack sentinel just to avoid including it in `properties`
```

File: lighthouse-core/lib/asset-saver.js [hunk 8]
```diff
@@ -100,8 +117,8 @@ async function saveArtifacts(artifacts, basePath) {
     fs.writeFileSync(`${basePath}/${passName}${devtoolsLogSuffix}`, log, 'utf8');
   }
 
-  // save everything else
-  const restArtifactsString = JSON.stringify(restArtifacts, null, 2);
+  // save everything else, using a replacer to serialize LHErrors in the artifacts.
+  const restArtifactsString = JSON.stringify(restArtifacts, stringifyReplacer, 2);
```
Comment (patrickhulce) [inline lighthouse-core/lib/asset-saver.js]: a wild non-null usage in stringify appears! 😮

File: lighthouse-core/lib/lh-error.js [hunk 9]
```diff
@@ -96,6 +103,74 @@ class LighthouseError extends Error {
...
+        code,
+        stack,
+      };
+    }
+
+    throw new Error('Invalid value for LHError stringification');
+  }
+
+  /**
+   * A JSON.parse reviver. If any value passed in is a serialized Error or
```
Comment (patrickhulce) [inline lighthouse-core/lib/lh-error.js]: worth linking to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#Using_the_reviver_parameter
?

TIL that it visits all the leaf nodes first before visiting parent objects :)

File: lighthouse-core/test/runner-test.js [hunk 10]
```diff
@@ -330,32 +330,40 @@ describe('Runner', () => {
...
-        assert.strictEqual(auditResult.scoreDisplayMode, 'error');
-        assert.ok(auditResult.errorMessage.includes(errorMessage));
-      });
+      const results = await Runner.run({}, {config});
+      const auditResult = results.lhr.audits['content-width'];
+      assert.strictEqual(auditResult.score, null);
+      assert.strictEqual(auditResult.scoreDisplayMode, 'error');
+      assert.ok(auditResult.errorMessage.includes(errorMessage));
+
+      rimraf.sync(resolvedPath);
```
Comment (patrickhulce) [inline lighthouse-core/test/runner-test.js]: should this be done in some sort of `after`?

File: lighthouse-core/lib/lh-error.js (hunk 6 above)
Comment (brendankenny) [reply lighthouse-core/lib/lh-error.js]: > we could leave the line in if we want to also nudge folks to look into the still expected ones that aren't LHError yet?

no, I think you're right. We should have some nudges but probably not here :)

File: lighthouse-core/lib/lh-error.js (hunk 9 above)
Comment (brendankenny) [reply lighthouse-core/lib/lh-error.js]: > TIL that it visits all the leaf nodes first before visiting parent objects :)

yeah, I was a little worried about performance, but it doesn't seem like a big hit, and all the "optimizations" I tried to add fast paths seemed to have no net effect (so maybe the biggest hit is just the overhead from having any replacer/reviver at all)

File: lighthouse-core/test/runner-test.js (hunk 10 above)
Comment (brendankenny) [reply lighthouse-core/test/runner-test.js]: > should this be done in some sort of `after`?

everything is sync before it, so it seemed ok. Are there tricky jest things to worry about or anything?

Comment (patrickhulce) [reply lighthouse-core/test/runner-test.js]: no I was just thinking if the assertions fail it won't be cleaned up, but this certainly won't be the only rough edge we have in that situation across the codebase
